src/utils/chunkColouring.ts

```diff
@@ -0,0 +1,83 @@
+import ExternalModule from '../ExternalModule';
+import Module from '../Module';
+import { randomUint8Array, Uint8ArrayXor } from './entryHashing';
+import { error } from './error';
+import { relative } from './path';
+
+export function assignChunkColouringHashes(
+	entryModules: Module[],
+	manualChunkModules: Record<string, Module[]>
+) {
+	let currentEntry: Module, currentEntryHash: Uint8Array;
+	let modulesVisitedForCurrentEntry: { [id: string]: boolean };
+	const handledEntryPoints: { [id: string]: boolean } = {};
+	const dynamicImports: Set<Module> = new Set();
+
+	const addCurrentEntryColourToModule = (module: Module) => {
+		if (currentEntry.chunkAlias) {
+			module.chunkAlias = currentEntry.chunkAlias;
+			module.entryPointsHash = currentEntryHash;
+		} else {
+			Uint8ArrayXor(module.entryPointsHash, currentEntryHash);
+		}
+
+		for (const dependency of module.dependencies) {
+			if (dependency instanceof ExternalModule || dependency.id in modulesVisitedForCurrentEntry) {
+				continue;
+			}
+			modulesVisitedForCurrentEntry[dependency.id] = true;
+			if (!handledEntryPoints[dependency.id] && !dependency.chunkAlias)
+				addCurrentEntryColourToModule(dependency);
+		}
+
+		for (const { resolution } of module.dynamicImports) {
+			if (
+				resolution instanceof Module &&
+				resolution.isDynamicEntryPoint &&
+				!resolution.chunkAlias
+			) {
+				dynamicImports.add(resolution);
+			}
+		}
+	};
+
+	if (manualChunkModules) {
+		for (const chunkName of Object.keys(manualChunkModules)) {
+			currentEntryHash = randomUint8Array(10);
+
+			for (currentEntry of manualChunkModules[chunkName]) {
+				if (currentEntry.chunkAlias) {
+					error({
+						code: 'INVALID_CHUNK',
+						message: `Cannot assign ${relative(
+							process.cwd(),
+							currentEntry.id
+						)} to the "${chunkName}" chunk as it is already in the "${
+							currentEntry.chunkAlias
+						}" chunk.
+Try defining "${chunkName}" first in the manualChunks definitions of the Rollup configuration.`
+					});
+				}
+				currentEntry.chunkAlias = chunkName;
+				modulesVisitedForCurrentEntry = { [currentEntry.id]: true };
+				addCurrentEntryColourToModule(currentEntry);
+			}
+		}
+	}
+
+	for (currentEntry of entryModules) {
+		handledEntryPoints[currentEntry.id] = true;
+		currentEntryHash = randomUint8Array(10);
+		modulesVisitedForCurrentEntry = { [currentEntry.id]: null };
+		addCurrentEntryColourToModule(currentEntry);
+	}
+
+	for (currentEntry of Array.from(dynamicImports)) {
+		if (handledEntryPoints[currentEntry.id]) {
+			continue;
+		}
+		currentEntryHash = randomUint8Array(10);
+		modulesVisitedForCurrentEntry = { [currentEntry.id]: null };
+		addCurrentEntryColourToModule(currentEntry);
+	}
```

Comment:
Ah I missed that the dynamic colouring is being done here.

Reply:
Ok, here's an example of the algorithm I was suggesting (`~>` means dynamic import, `->` means static import):

```
E1
~> D1 -> F -> T
-> S -> T      
-> R

E2
~> D2 -> G -> R
-> S -> T      
-> R

E3
~> D3 -> H -> S->T
-> S -> T      
-> R
```

The current colouring algorithm will result in no optimization groups at all in the above, as every module has a unique colouring.

On the other hand, we could get a more optimal chunking since we know that S, T and R are always loaded together _despite_ the dynamic imports being different colours:

```
E1
~> D1 -> F -> STR
-> STR

E2
~> D2 -> G -> STR
-> STR  

E3
~> D3 -> H -> STR
-> STR
```

The algorithm is, for each dynamic import, to determine the intersection of the static graph of all of its importers. So for each of D1, D2 and D3 this is STR. Then when doing colouring for those dynamic imports, to skip these shared items when it comes to colouring, as the dynamic import isn't affecting their colours as they are already available.

This algorithm is also the general case of resolving the issue we have with the test case I provided:

```js
Main -> dep1 -> dep2 ~> dep1
```

dep1 has parent importer static intersection Main, dep1, dep2. Therefore the colouring of dep1 doesn't recolour itself and as a result we get a single chunk.

Perhaps for another day, but glad to be able to verify the logic here. Let me know if that all sounds correct too please!

Reply:
Note also here that dynamic import facades can be rendered inline:

```js
import('./static-chunk').then(x => ({ get some () { return x.a }, get export () { return x.b }, get names () { return x.c } }));
```

Reply:
Thanks a lot, this really helped understand it! Definitely worth adding this in the future!

> Note also here that dynamic import facades can be rendered inline

Just a minor question, you are using getters instead of `Object.freeze`, any particular advantage or disadvantage to doing this? Or is `Object.freeze` semantically wrong here?

Otherwise this all really makes sense, I'll add it to the other issue.

Reply:
I assume we could also just use the existing namespace import algorithm, i.e. add that as another import and then just pluck that from the dynamic import result?